[2022.10.19] feat: firebase auth를 사용한 로그인 구현

rktnsinger opened the pull request:
## 작업 사항
- 로그인 후 firebase session cookie(httponly cookie)를 사용하여 서버에 인증된 요청 전송
- localStorage에 firebase에서 제공한 google access token으로 부터 생성되는 jwt token을 저장하여 클라이언트 측 인증상태 지속
- 로그인 관련 api 요청 작성
- 콘솔페이지엔 인증된 유저만 접근하도록 ProtectedRoute 래퍼 제작
- 로그인, 로그아웃 관련 함수 커스텀 훅으로 제작

기타 변경사항
- 의존성 변경: firebase, dotenv 추가됨
- eslint가 commit시 자동으로 컨벤션을 수정하는 동작을 비활성화함

## 잘 봐주세요
- 작업 초반에는 로그인, 로그아웃 관련로직을 하나의 커스텀 훅으로 제작하여, 내부에서 state를 관리하는 훅으로 제작하고 싶었으나 시간이 지체되어 일반 유틸함수 정도의 기능으로만 구현이 되었습니다. 이에 useNavigate 훅을 사용하지 않고 다른 redirect 방법을 취한다면, 굳이 커스텀 훅으로써 사용할 필요는 없어보이기도 합니다(유틸 함수로 충분). 이부분 의견 부탁드립니다.
- 페이지 레이아웃 작업 이외에 FE쪽에 비즈니스 로직이 들어가는 첫 작업물인지라, import 순서나 컨벤션 등이 잘 맞는지 꼼꼼히 리뷰를 부탁드립니다.
- api 요청에 반복되는 로직들을 추후 커스텀 훅으로 추출하여 재사용 할 수 있을 것 같습니다. 이는 에러처리와도 관련이 깊어 비교적 작업량이 적은 에러페이지 제작 task 카드 수행시 함께 작업해보면 어떨까 합니다.(에러 코드, 메시지 등이 전달되도록 연결 필요) 이부분도 의견 부탁드립니다.

## PR 전 확인사항
- [x] 가장 최신 브랜치를 pull했습니다.
- [x] base 브랜치명을 확인했습니다.
- [x] 코드 컨벤션을 모두 지켰습니다.
- [x] 적절한 라벨이 있습니다.
- [x] assignee가 있습니다.

## 비고
- 작성된 authApi 중 getUser api는 /auth/user로 GET 요청을 보내 session cookie를 기반으로 db에서 사용자 정보를 받아올 수 있는 요청인데, 처음에는 테스트 용으로 작성한 로직이나 추후 각 페이지에 대한 연결 작업시 필요할 가능성이 있을 것으로 판단되어 우선 삭제하지 않고 남겨두었습니다. 혹 지금은 삭제하고 추후 다시 작성하는 것이 좋겠다면 리뷰 부탁드릴께요~!

Commit:
feat: firebase auth를 사용한 로그인 구현

- 로그인 후 firebase session cookie를 사용하여 서버에 인증
- localStorage에 jwt token을 저장하여 클라이언트 측 인증상태 지속
- 로그인 관련 api 요청 작
- 콘솔페이지엔 인증된 유저만 접근하도록 ProtectedRoute 래퍼 제작

기타 변경사항
- 의존성 변경: firebase, dotenv 추가됨

Yanguk commented on package.json:
```diff
@@ -8,6 +8,8 @@
     "@testing-library/react": "^13.4.0",
     "@testing-library/user-event": "^13.5.0",
     "prop-types": "^15.8.1",
+    "dotenv": "^16.0.3",
```
`dotenv`가 사용되는 곳이 있을까요?
CRA에서 제공되는 REACT_APP 으로는 안되는 부분이 있었을까요?

rktnsinger replied:
헉 제가 착각했어요! 수정하겠습니다 👍

Yanguk commented on package.json:
```diff
@@ -35,7 +37,7 @@
   },
   "lint-staged": {
-    "*.js": "eslint --fix"
+    "*.js": "eslint"
```
강제 수정을 없애는거에 다들 동의하시는 거죠?

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+import config from "../config/config";
```
```suggestion
import config from "../config";
```
네이밍이 중복됩니다. 파일명을 index로 변경하는게 좋아보여요

rktnsinger replied:
반영했습니다!

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+
+  try {
+    const response = await fetch(
+      `${config.REACT_APP_REQUEST_URL}/auth/users`,
```
저희쪽 서버를 환경변수로 등록하신건가요??
그렇다면 좋은 아이디어인 것 같습니다. 서버를 따로 배포하게되면 url의 수정이 일어날수가있으니까요 👍

rktnsinger replied:
혹시 변수명 `REQUEST_URL` 괜찮나요? `DEFAULT_SERVER_URL` 이런것두 고민해보긴 했는데.. 😑

Yanguk replied:
넵 좋아요 👍

ssong-yi replied:
```suggestion
      `${config.REACT_APP_SERVER_URL}/auth/users`,
```
`SERVER_URL` 좋습니다!

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+  } catch (error) {
+    console.log(error);
+  }
+};
```
폴더 명의 authApi 가 자연스럽지 않은 것 같아요.
api는 아니고 요청을 보내는 형식이니 `fetchAuth` 같이 서버에 요청에 날리는 뜻이 포함되는게 더 자연스러워 보인다고 생각합니다

rktnsinger replied:
의견 감사해요~! 통상적으로 사용되는 컨벤션으로 지은 이름이긴 한데, 논의해보고 적절하게 변경하면 좋겠네요 ㅎㅎ

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+import config from "../config/config";
+
+export const getAuthUser = async () => {
+  const request = {
```
```suggestion
try {
  const request = {
```
트라이 를 최상단에 선언해서 하는게 코드가 더 깔끔하게 보일 것 같아요

rktnsinger replied:
아래 논의한 대로 해당 함수를 사용하는 컴포넌트에서 에러 처리하도록 여기서는 `try...catch` 제거하겠습니다!

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+
+    return data;
+  } catch (error) {
+    return console.log(error);
```
여기서 error를 잡고 console만 남기고 끝내버리면 error가 뿜어 나와도 로직은 흘려가고 콘솔만 찍히게 됩니다.
제 생각에는 여기서 컨트롤 하기는 힘들 것 같고 여기서는 에러를 뿜게 냅두고 리액트 컴포넌트 안에서 에러를 캐치하여서 error페이지로 보내는 방식이 어떨까 제안 드립니다

rktnsinger replied:
맞네요! 백엔드 작업 동시에 하다보니 조금 혼동이 왔는데, 여기서는 냅두고 이 api를 사용하는 함수에서 처리하면 될 것 같습니다 👍  바로 반영할께요!

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+  }
+};
+
+export const findOrCreateUser = async (userData, token) => {
```
```suggestion
export const fetchRegisteUser = async (userData, token) => {
```
적절한 함수명이 뭐가 있을까요??
여기서 유저를 찾거나 만드는 행위는 하지않아서 네이밍이 부자연스러워 보입니다

rktnsinger replied:
음 고민해보니 `fetch`라는 단어가 들어가면 추후 `fetch`관련 유틸함수나 커스텀 훅이 제작되었을때 헷갈릴 수도 있을 것 같아서, 여기서 `getUser`로 하고, 이미 세션쿠키가 발급되어 인증된 사용자의 정보를 가져오는 함수는 `getAuthUser`로 하면 어떨까요??

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+  }
+};
+
+export const fetchLogout = async () => {
```
👍  여기서는 fetch를 넣어주셨네요

rktnsinger replied:
ㅠㅠ 여기 도저히 넣을 이름이 없어서 고민하다가 일단 `fetch`로 했는데, 통일감이 없으니 아예 `fetch`로 다 맞추거나 `getLogout`, `tryLogout` 등 뭔가 다른 이름을 생각해봐야할 듯 합니다!

아니면 그냥 `logout`이나 `signout`은 어떨까요..?

ssong-yi replied:
`logOut` 괜찮은 것 같습니다!

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+  try {
+    await fetch(`${config.REACT_APP_REQUEST_URL}/auth/logout`, request);
+  } catch (error) {
+    console.log(error);
```
여기도 에러를 잡았으면 그에따른 행동을 해야할 것 같은데 뭐가 좋을지 의논해보면 좋겠어요
그런데! 밑에서 에러핸들링을 하셨으니 여기는 에러를 뿜어내는 쪽으로 하셔야할 것 같네요

Yanguk commented on src/app/App.js:
```diff
@@ -19,11 +20,13 @@ export default function App() {
-          <Route path=":projectId" element={<ProjectDetailPage />}>
-            <Route path="channels" element={<ChannelsPage />}>
-              <Route path=":channelId" element={<ChannelDetailPage />} />
+      <Route element={<ProtectedRoute />}>
```
라우터로 접근 권한을 거는 방식 아주 좋은 것 같아요 👍  같이 학습해 나가네요

ssong-yi replied:
저도 덕분에 새로운 방법 배웠네요!

Yanguk commented on src/api/authApi.js:
```diff
@@ -0,0 +1,57 @@
+import config from "../config/config";
+
+export const getAuthUser = async () => {
```
다음사람을 위해 미리 만들어두셨군요 좋습니다 👍

Yanguk commented on src/hooks/useLogin.js:
```diff
@@ -0,0 +1,31 @@
+import { findOrCreateUser } from "../api/authApi";
+import { signInWithGoogle } from "../config/firebaseConfig";
+
+export default function useLogin() {
```
훅 으로 만들어 주셨네요 👍  다른사람도 재사용하기 쉽겠네요

Yanguk commented on src/hooks/useLogin.js:
```diff
@@ -0,0 +1,31 @@
+      const token = await authData.user.getIdToken();
+
+      const { email, displayName, photoURL } = authData.user;
+      const userData = await findOrCreateUser(
```
이부분에서도 에러를 잡아주셨군요!
그렇다면 위에 남긴 코멘트처럼
`findOrCreateUser` 함수에서 에러날시 에러를 뿜어내는 쪽으로 해주셔야 여기서 에러가 같이 처리가 될 것 같아요

Yanguk commented on src/layout/HomeLayOut/Navbar.js:
```diff
@@ -31,7 +32,7 @@ const Container = styled.nav`
   align-items: center;
   position: fixed;
-  width: 100%;
+  width: 100vw;
```
반응형 고려 좋습니다 👍

Yanguk commented on src/hooks/useLogin.js:
```diff
@@ -0,0 +1,31 @@
+      );
+
+      if (userData) {
+        localStorage.setItem("authToken", token);
```
로컬스토리지에 저장하는 이름인 `"authToken"`을 상수화 하면 어떨까요? 그러면 서버 작업시 합맞추기가 수월할 듯 하네요
그리고 서버에서의 verifyToken 미들웨어를 로컬스토리지에서 꺼내는 형식으로 추후 변경 해야겠네요

rktnsinger replied:
혹시 `httpOnly` 옵션을 끄고 세션 쿠키를 기반으로 클라이언트에서도 사용자를 인증하면 어떨까요?? 다시 생각해보니 굳이 이원화해서 저장할 필요가 없는 것 같습니다..!

ssong-yi replied:
보안을 생각하면 `httpOnly`를 `true`로 주는 게 맞는 것 같은데, 장단점을 비교해보면 좋을 것 같네요!

rktnsinger replied:
논의대로 이 부분은 쿠키를 기반으로 인증하는 것으로 변경하겠습니다!

rktnsinger commented on src/config/config.js:
```diff
@@ -0,0 +1,13 @@
+const envKeys = {
```
env.sample 업데이트하세요

Yanguk reviewed (commented):
로그인 하시느라 고생하셨어요! 리뷰남긴 코멘트 가지고 논의하고 수정되면은 어프로브 진행하겠습니다.
package.json 패치도 버전업 해주세요~

ssong-yi reviewed (commented):
형욱님 고생많으셨어요!

리뷰 확인부탁드립니다~

Commit:
refactor: 코드리뷰 반영
- 클라이언트 측 인증상태 지속의 기준을 쿠키를 인식하는 것으로 변경
- 비동기 api요청에 대한 에러처리 로직 수정
- 그밖의 자잘한 수정 및 버전 업데이트

Yanguk commented on src/hooks/useLogin.js:
```diff
@@ -1,28 +1,23 @@
-        localStorage.setItem("authToken", token);
-
-        navigate("/console/projects", { replace: true });
+        navigate("/console/projects", { rep: true });
```
```suggestion
        navigate("/console/projects", { replace: true });
```
오타일까요?

rktnsinger replied:
부끄럽네요.. 죄송합니다 바로 수정하겠습니다!

Yanguk commented on src/pages/ProtectedRoute/index.js:
```diff
@@ -2,7 +2,23 @@ import React from "react";
+    const decodedCookies = decodeURIComponent(document.cookie);
+    const cookies = decodedCookies.split("; ");
+
+    let matchedCookie = "";
```
```suggestion
    let matchedCookie = false;
```

matchedCookie는 boolean 값 인가요???? 타입을 맞추는게 좋아보일듯 하네요

Yanguk commented on src/pages/ProtectedRoute/index.js:
```diff
@@ -2,7 +2,23 @@ import React from "react";
+
+    cookies.forEach((value) => {
+      if (value.indexOf(name) === 0) {
+        matchedCookie = value.substring(name.length);
```
```suggestion
        matchedCookie = true;
```
불린값으로써 활용할 의도가 맞을까요?

ssong-yi replied:
`matchedCookie`가 `getCookie`의 리턴값이고, `isAuthUser`에 들어가서 `true`, `false` 판단시에만 이용하는 것으로 보여서 `Boolean` 타입이 적절할 것 같습니다!

rktnsinger replied:
넵 반영하였습니다!!

Yanguk reviewed (commented):
고생하셨습니다! 코드리뷰 봐주세용

Commit:
refactor: 코드리뷰 반영(2차)
- checkSessionCookie 함수 내 변수명, 값 수정
- await 구문 누락부분 수정

ssong-yi reviewed (approved):
수정확인했습니다!

Yanguk reviewed (approved):
반영사항 확인하였습니다~ 고생하셨어요 👍
